rs/messaging/src/routing/stream_handler/tests.rs

```diff
@@ -3468,281 +2859,603 @@ fn process_stream_slices_canister_migration_in_both_subnets_success() {
 }
 
-fn generate_stream_slice(config: StreamSliceConfig) -> StreamSlice {
-    let stream = generate_stream(
-        MessageConfig {
-            sender: *REMOTE_CANISTER,
-            receiver: *LOCAL_CANISTER,
-            begin: config.messages_begin,
-            count: config.message_count,
-        },
-        SignalConfig {
-            end: config.signals_end,
-            reject_signals: config.reject_signals,
-        },
-    );
-    let mut slice: StreamSlice = stream.into();
-    slice.header_mut().set_begin(config.header_begin.into());
-    if let Some(end) = config.header_end {
-        slice.header_mut().set_end(end.into());
+/// Returns an iterator over messages in a stream over the `stream_index_range`.
+///
+/// Panics if any of the messages in `stream_index_range` does not exist.
+fn messages_in_stream(
+    opt_stream: Option<&Stream>,
+    stream_index_range: std::ops::RangeInclusive<u64>,
+) -> impl Iterator<Item = &RequestOrResponse> {
+    match opt_stream {
+        // Not the `unwrap()`s here ensure that the code panics when the whole range was not
+        // available rather than just silently terminating the iterator prematurely. This behaviour
+        // is important to ensure we actually test as intended.
+        Some(stream) => stream_index_range
+            .map(|stream_index| stream.messages().get(stream_index.into()).unwrap()),
+        None => unreachable!(),
     }
-    slice.header_mut().set_flags(config.flags);
-    slice
 }
 
-fn get_request_at(queue: &StreamIndexedQueue<RequestOrResponse>, stream_index: u64) -> &Request {
-    match queue.get(stream_index.into()) {
-        Some(RequestOrResponse::Request(request)) => request,
-        _ => unreachable!(),
+/// Returns an iterator over messages in a stream slice over the `stream_index_range`.
+///
+/// Panics if any of the messages in `stream_index_range` does not exist.
+fn messages_in_slice(
+    opt_slice: Option<&StreamSlice>,
+    stream_index_range: std::ops::RangeInclusive<u64>,
+) -> impl Iterator<Item = &RequestOrResponse> {
+    match opt_slice {
+        // Not the `unwrap()`s here ensure that the code panics when the whole range was not
+        // available rather than just silently terminating the iterator prematurely. This behaviour
+        // is important to ensure we actually test as intended.
+        Some(slice) => stream_index_range
+            .map(|stream_index| slice.messages().unwrap().get(stream_index.into()).unwrap()),
+        None => unreachable!(),
     }
 }
 
-fn get_response_at(queue: &StreamIndexedQueue<RequestOrResponse>, stream_index: u64) -> &Response {
-    match queue.get(stream_index.into()) {
-        Some(RequestOrResponse::Response(response)) => response,
-        _ => unreachable!(),
+/// Pushes the messages yielded by `iter` into the `state`.
+fn push_inputs<'a>(
+    state: &mut ReplicatedState,
+    iter: impl Iterator<Item = &'a RequestOrResponse>,
+) -> Result<(), (StateError, RequestOrResponse)> {
```

Comment:
I believe we always just `unwrap()` the output of `push_inputs()`, so we may as well do it here and not return anything.

I believe we do actually look at the output of `push_input()` once, but even in that case, I would rather have `push_input()` be infallible and either have a separate `try_push_input()` function or (better) simply have the one caller use `state.push_input()` directly.

Reply:
Removed the `unwrap()`. The one use where it used the error now just uses the naked state.